### PR TITLE
README Rewrite

### PR DESCRIPTION
**Pull request by ShayneP**
The README has been a bit of a mess for a while, so I rewrote it.

Highlights :
- Added a TOC
- Moved the breaking changes into their own section that is at the end
- Added a section on Bulk
- Split up "Getting Started" into more logical pieces, and have separate paths for GraphQL vs REST
- Highlighted the importance of GraphQL for clients

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,134 +5,61 @@ Shopify API
...
+
+The REST API is implemented as JSON over HTTP using all four verbs (GET/POST/PUT/DELETE). Each resource, like Order, Product, or Collection, has a distinct URL and is manipulated in isolation. In other words, we’ve tried to make the API follow the REST principles as much as possible.
+
+- [Shopify API](#shopify-api)
+- [Usage](#usage)
+  * [Requirements](#requirements)
+    + [Ruby version](#ruby-version)
+  * [Installation](#installation)
+  * [Getting Started](#getting-started)
+    + [1) Create an app](#1--create-an-app)
```
Comment (tiffanytse) [inline README.md]: All of these bookmarks don't link properly, there's an extra `-` after the number, should be `#1-create-an-app`

File: README.md [hunk 2]
```diff
@@ -5,134 +5,61 @@ Shopify API
...
+    + [3) Requesting access from a shop](#3--requesting-access-from-a-shop)
+    + [4) Trading your `code` for an access token.](#4--trading-your--code--for-an-access-token)
+    + [5) Activating the session](#5--activating-the-session)
+    + [6A) Making requests to the GraphQL API](#6a--making-requests-to-the-graphql-api)
+    + [6B) Making requests to the REST API](#6b--making-requests-to-the-rest-api)
+  * [Console](#console)
+  * [Thread safety](#thread-safety)
+  * [Bulk Operations](#bulk-operations)
+    + [Example](#example)
+      - [1) Start the bulk operation](#1--start-the-bulk-operation)
```
Comment (tiffanytse) [inline README.md]: Same here

File: README.md [hunk 3]
```diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
...
 
-1. First create a new application in either the partners admin or your store admin. For a private App you'll need the API_KEY and the PASSWORD otherwise you'll need the API_KEY and SHARED_SECRET.
+**Public apps** can be installed on many stores, and can be added to the Shopify App Store to generate revenue for the developer. 
 
-   If you're not sure how to create a new application in the partner/store admin and/or if you're not sure how to generate the required credentials, you can [read the related shopify docs](https://docs.shopify.com/api/guides/api-credentials) on the same.
+For a private app, you'll need the API_KEY and the PASSWORD; otherwise, you'll need the API_KEY and SHARED_SECRET.
 
-2. For a private App you just need to set the base site url as follows:
+   If you're not sure how to create a new application in the partner/store admin or if you're not sure how to generate the required credentials, you can [read the related Shopify docs](https://docs.shopify.com/api/guides/api-credentials) on the same.
```
Comment (tiffanytse) [inline README.md]: Update url to shopify.dev

File: README.md [hunk 4]
```diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
...
+For a private App you just need to set the base site url as follows:
 
    ```ruby
    shop_url = "https://#{API_KEY}:#{PASSWORD}@#{SHOP_NAME}.myshopify.com"
    ShopifyAPI::Base.site = shop_url
    ShopifyAPI::Base.api_version = '<version_name>' # find the latest stable api_version [here](https://help.shopify.com/api/versioning)
    ```
 
-   That's it, you're done, skip to step 6 and start using the API!
+   That's it; you're done! Next, skip to step 6 and start using the API!
```
Comment (tiffanytse) [inline README.md]: Don't think this needs to change to a semicolon here?

File: README.md [hunk 5]
```diff
@@ -182,59 +139,41 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
...
-
-   Before you proceed, make sure your application performs the following security checks. If any of the checks fails, your application must reject the request with an error, and must not proceed further.
-
-   * Ensure the provided ``state`` is the same one that your application provided to Shopify during Step 3.
-   * Ensure the provided hmac is valid. The hmac is signed by Shopify as explained below, in the Verification section.
-   * Ensure the provided hostname parameter is a valid hostname, ends with myshopify.com, and does not contain characters other than letters (a-z), numbers (0-9), dots, and hyphens.
-
-   If all security checks pass, the authorization code can be exchanged once for a permanent access token. The exchange is made with a request to the shop.
+   This method will save the token to the session object and return it. All fields returned by Shopify, other than the access token itself, are stored in the session's `extra` attribute. For a list of all fields returned by Shopify, read [our OAuth documentation](https://help.shopify.com/api/guides/authentication/oauth#confirming-installation).
```
Comment (tiffanytse) [inline README.md]: url update here too (https://shopify.dev/tutorials/authenticate-with-oauth#confirming-installation)

File: README.md (hunk 3 above)
Comment (tiffanytse) [reply README.md]: https://shopify.dev/tutorials/authenticate-a-private-app-with-shopify-admin

File: README.md [hunk 6]
```diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
...
+   If you're not sure how to create a new application in the partner/store admin or if you're not sure how to generate the required credentials, you can [read the related Shopify docs](https://docs.shopify.com/api/guides/api-credentials) on the same.
+
+### 2A) Private Apps
+
+For a private App you just need to set the base site url as follows:
 
    ```ruby
    shop_url = "https://#{API_KEY}:#{PASSWORD}@#{SHOP_NAME}.myshopify.com"
    ShopifyAPI::Base.site = shop_url
    ShopifyAPI::Base.api_version = '<version_name>' # find the latest stable api_version [here](https://help.shopify.com/api/versioning)
```
Comment (tiffanytse) [inline README.md]: url update

File: README.md [hunk 7]
```diff
@@ -182,59 +139,41 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
 
    * ``client_id`` – Required – The API key for your app
    * ``scope`` – Required – The list of required scopes (explained here: https://help.shopify.com/api/guides/authentication/oauth#scopes)
```
Comment (tiffanytse) [inline README.md]: url update

Comment (tiffanytse) [reply README.md]: https://shopify.dev/tutorials/authenticate-with-oauth#scopes

File: README.md [hunk 8]
```diff
@@ -247,24 +186,58 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
...
    ```
 
-6. Now you're ready to make authorized API requests to your shop! Data is returned as ActiveResource instances:
+### 6A) Making requests to the GraphQL API
+
+The GraphQL API is the recommended way to consume the Shopify API. It is more fully-featured than REST, more performant, and future-proof. Whenever possible, GraphQL should be used to consume the Shopify API.
+
+###### Note: the GraphQL client has improved and changed in version 9.0. See the [client documentation](docs/graphql.md) for full usage details and a [migration guide](docs/graphql.md#migration-guide).
+
+This library also supports Shopify's [GraphQL Admin API](https://help.shopify.com/api/graphql-admin-api)
```
Comment (tiffanytse) [inline README.md]: https://shopify.dev/docs/admin-api/graphql/reference

File: README.md [hunk 9]
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
    shopify-api help
    ```
 
-## GraphQL
+## Thread safety
 
-Note: the GraphQL client has improved and changed in version 9.0. See the [client documentation](docs/graphql.md)
-for full usage details and a [migration guide](docs/graphql.md#migration-guide).
+ActiveResource is threadsafe as of version 4.1 (which works with Rails 4.x and above).
```
Comment (tiffanytse) [inline README.md]: should this be marked up as code? `4.x`

File: README.md [hunk 10]
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
...
-Page based pagination
 ```ruby
-page = 1
-products = ShopifyAPI::Product.find(:all, params: { limit: 50, page: page })
-process_products(products)
-while(products.count == 50)
-  page += 1
-  products = ShopifyAPI::Product.find(:all, params: { limit: 50, page: page })
-  process_products(products)
+# Efficient: reads the file a single line at a time
```
Comment (tiffanytse) [inline README.md]: Nice!

File: README.md [hunk 11]
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
...
+
+jsonl.each_line do |line|
+  JSON.parse(line)
 end
 ```
 
-Page based pagination will be deprecated in the `2019-10` API version, in favor of the second method of pagination:
+## Pagination
+
+Shopify uses [Relative cursor-based pagination](https://help.shopify.com/en/api/guides/paginated-rest-results) to provide more than a single page of results.
```
Comment (tiffanytse) [inline README.md]: https://shopify.dev/tutorials/make-paginated-requests-to-rest-admin-api

File: README.md [hunk 12]
```diff
@@ -437,11 +620,11 @@ or you can even use our automated rake task for docker:
 bundle exec rake docker
 ```
 
-## Additional Resources
+# Additional Resources
 
 * [API Reference](https://help.shopify.com/api/reference)
```
Comment (tiffanytse) [inline README.md]: https://shopify.dev/docs/admin-api/rest/reference

File: README.md (hunk 3 above)
Comment (ShayneP) [reply README.md]: This isn't the same guide, I'll have to go back and rewrite the section to account for the change.

File: README.md (hunk 4 above)
Comment (ShayneP) [reply README.md]: They're both independent clauses, so a comma creates a comma splice. We could put a period instead, but I figure the flow is a bit better with the semicolon.

File: README.md (hunk 1 above)
Comment (ShayneP) [reply README.md]: Good catch, forgot about that. Thanks!

**Comment by ShayneP** [comment]
Added links to the new docs so that they don't have to redirect, rewrote the authentication section to account for the new tutorial format, changed forum links to the new forums, and added GraphQL Reference link at the end!

File: README.md (hunk 3 above)
Comment (ShayneP) [reply README.md]: Done!

**Pull request merged**